### PR TITLE
feat: add metro transformer

### PR DESCRIPTION
**Pull request by vonovak**
# Description

This adds the equivalent of webpack loader for metro bundler used in react native apps.

I'm targeting the main branch but could easily switch to `next` if desirable.

## Types of changes

[//]: # (What types of changes does your code introduce to Lingui?)
[//]: # (_Put an `x` in the boxes that apply_)

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Documentation update
- [ ] Examples update

Fixes # (issue)

## Checklist

[//]: # (_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._)

- [x] I have read the [CONTRIBUTING](https://github.com/lingui/js-lingui/blob/main/CONTRIBUTING.md) and [CODE_OF_CONDUCT](https://github.com/lingui/js-lingui/blob/main/CODE_OF_CONDUCT.md) docs
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added the necessary documentation (if appropriate)

### CODE REVIEW

File: packages/metro-transformer/package.json [hunk 1]
```diff
@@ -0,0 +1,68 @@
+{
+  "name": "@lingui/metro-transformer",
+  "version": "0.1.0",
```
Comment (andrii-bodnar) [inline packages/metro-transformer/package.json:3]: Please align with the current Lingui version and add publish config:

```
"publishConfig": {
  "access": "public"
},
```

This allows the package to be published by CI for the first time

**Comment by andrii-bodnar** [comment]
@vonovak thank you!

Looks very nice to me, just left a small comment in the `package.json` file.

I was thinking about including it in the v5 release but realized that there is no reason to postpone it, and it would probably be better to ship it in v4 so that users can already use it. I hope it will not cause a lot of conflicts when merging `main` into `next`.

File: packages/metro-transformer/src/expo/index.ts [hunk 2]
```diff
@@ -0,0 +1,5 @@
+import { createLinguiMetroTransformer } from "../metroTransformer"
+
+const expoTransformer = require("@expo/metro-config/babel-transformer")
+
+export const transform = createLinguiMetroTransformer(expoTransformer)
```
[the remark this answers is not in the document]
Comment (vonovak) [reply packages/metro-transformer/src/expo/index.ts]: no, changed to import :)

**Comment by vonovak** [comment]
> I was thinking about including it in the v5 release but realized that there is no reason to postpone it, and it would probably be better to ship it in v4 so that users can already use it. I hope it will not cause a lot of conflicts when merging `main` into `next`.

@andrii-bodnar I don't mind merging into `next` and shipping with v5, but I think you're right. It can be mentioned in v5 release notes as one of the new features anyway. I don't think it should cause conflicts, maybe just in yarn.lock file.

File: packages/metro-transformer/src/expo/index.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (vonovak) [reply packages/metro-transformer/src/expo/index.ts:5]: actually, changed it back because rollup was complaining about `implicit dependency`. 

`"@expo/metro-config/babel-transformer"` is an optional peer dependency.

File: packages/metro-transformer/package.json [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
```
Comment (andrii-bodnar) [inline packages/metro-transformer/package.json:3]:
```suggestion
  "version": "4.11.4",
```

File: packages/metro-transformer/package.json [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+  },
+  "repository": {
+    "type": "git",
+    "url": "https://github.com/lingui/js-lingui.git"
+  },
+  "bugs": {
+    "url": "https://github.com/lingui/js-lingui/issues"
+  },
+  "engines": {
+    "node": ">=18.0.0"
```
Comment (andrii-bodnar) [inline packages/metro-transformer/package.json:50]: Currently, all the other packages require Node 16. I guess it would be better to do the same here if there are no blockers

File: packages/metro-transformer/src/expo/index.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (timofei-iatsenko) [reply packages/metro-transformer/src/expo/index.ts:5]: If it's a dependency it should be explicitly stated in the one of the dependencies in package.json, i see only `metro` stated as peer dependency, but not a `@expo/metro-config` 

That might cause bugs due to package managers will be not aware that this package is using that dependency.

Also, what you mean by "optional peer dependency"?   I don't see any branching here, for a case where this dependecny would not be resolved, as well as [peerDependeciesMeta](https://docs.npmjs.com/cli/v10/configuring-npm/package-json#peerdependenciesmeta)

File: packages/metro-transformer/src/react-native/index.ts [hunk 5]
```diff
@@ -0,0 +1,11 @@
+import { createLinguiMetroTransformer } from "../metroTransformer"
+
+const reactNativeTransformer = (() => {
+  try {
+    return require("@react-native/metro-babel-transformer")
+  } catch (error) {
+    return require("metro-react-native-babel-transformer")
+  }
```
Comment (timofei-iatsenko) [inline packages/metro-transformer/src/react-native/index.ts:8]: these two also should be in package json as optional peers

File: website/docs/ref/metro-transformer.mdx [hunk 6]
```diff
@@ -0,0 +1,101 @@
+# Metro transformer
+
+[Metro bundler](https://metrobundler.dev/) is a JavaScript bundler used in React Native apps. This package offers an alternative to the [`lingui compile`](/ref/cli#compile) command: a transformer that enables Metro to compile `.po` files on the fly. The transformer enables you to `import` `.po` files directly, instead of running `lingui compile` and `import`ing the resulting JavaScript (or TypeScript) files.
+
+## Installation
+
+> Only Expo SDK 49 and React Native v0.72.1 or newer are supported.
```
Comment (timofei-iatsenko) [inline website/docs/ref/metro-transformer.mdx:7]: I think this could be declared as constraints in peerDependecies as well

**Comment by andrii-bodnar** [comment]
@vonovak I plan to release 4.12.0 soon, it would be great to include this PR in it.

By the way, I was looking for some topics to improve the Lingui blog. This seems to be a great one. Would you like to write a little blog post about the Metro transformer?

File: packages/metro-transformer/src/expo/index.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (vonovak) [reply packages/metro-transformer/src/expo/index.ts:5]: done

File: packages/metro-transformer/src/react-native/index.ts (hunk 5 above)
Comment (vonovak) [reply packages/metro-transformer/src/react-native/index.ts:8]: done

File: website/docs/ref/metro-transformer.mdx (hunk 6 above)
Comment (vonovak) [reply website/docs/ref/metro-transformer.mdx:7]: done

File: packages/metro-transformer/src/expo/index.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (timofei-iatsenko) [reply packages/metro-transformer/src/expo/index.ts:5]: Thanks, and now import could be used and rollup will not complain

File: packages/metro-transformer/package.json (hunk 3 above)
Comment (vonovak) [reply packages/metro-transformer/package.json]: done

File: packages/metro-transformer/package.json (hunk 4 above)
Comment (vonovak) [reply packages/metro-transformer/package.json]: done

File: packages/metro-transformer/src/react-native/index.ts (hunk 5 above)
Comment (vonovak) [reply packages/metro-transformer/src/react-native/index.ts]: I ditched the `metro-react-native-babel-transformer` support because it's old and used import for `@react-native/metro-babel-transformer`

**Review by timofei-iatsenko — APPROVED** [review]

**Review by andrii-bodnar — APPROVED** [review]

**Comment by efstathiosntonas** [comment]
Hey guys, how to configure this in case there's another `babelTransformPath` like:

```ts
const { getDefaultConfig, mergeConfig } = require('@react-native/metro-config');

const defaultConfig = getDefaultConfig(__dirname);
const { assetExts, sourceExts } = defaultConfig.resolver;

metroConfig = (() => {
  return {
    transformer: {
      babelTransformerPath: require.resolve('react-native-svg-transformer/react-native')
    },
    resolver: {
      assetExts: [...assetExts.filter((ext) => ext !== 'svg'), 'bin', 'lottie'],
      sourceExts: [...sourceExts, 'svg']
    }
  };
})();

module.exports = mergeConfig(defaultConfig, metroConfig);
```

**Comment by efstathiosntonas** [comment]
nvm, figured it out: 

```ts
const { getDefaultConfig, mergeConfig } = require('@react-native/metro-config');

const defaultConfig = getDefaultConfig(__dirname);
const { assetExts, sourceExts } = defaultConfig.resolver;

metroConfig = (() => {
  return {
    transformer: {
      babelTransformerPath: require.resolve('@lingui/metro-transformer/react-native')
    },
    resolver: {
      assetExts: [...assetExts.filter((ext) => ext !== 'svg'), 'bin', 'lottie'],
      sourceExts: [...sourceExts, 'svg']
    }
  };
})();

linguiConfig = (() => {
  return {
    transformer: {
      babelTransformerPath: require.resolve('@lingui/metro-transformer/react-native')
    },
    resolver: {
      sourceExts: [...sourceExts, 'po', 'pot']
    }
  };
})();

module.exports = mergeConfig(defaultConfig, metroConfig, linguiConfig);
```

I think this should go into the [docs](https://lingui.dev/ref/metro-transformer), could save some time for the fellow devs.

**Comment by vonovak** [comment]
Hello,
@efstathiosntonas thanks for suggestion; I'm afraid that what you provided isn't going to work - it's more likely that one config will override the other.

I'd recommend this: https://stackoverflow.com/a/57660231/2070942

**Comment by efstathiosntonas** [comment]
@vonovak thanks for the heads up, I’ve already tried the split version based on file extensions but I few errors showed up. 

Let me try again and try to fix it since it seems to be the only solution. Will leave a comment with my findings.

**Comment by efstathiosntonas** [comment]
@vonovak I ended up with this and works as a charm for both svg imports and lingui po(t).

`customTransformer.js`: 

```js
const upstreamTransformer = require('metro-babel-transformer');
const linguiTransformer = require('@lingui/metro-transformer/react-native');
const svgTransformer = require('react-native-svg-transformer');

module.exports.transform = function ({ src, filename, options }) {
  if (filename.endsWith('.po' || filename.endsWith('pot'))) {
    return linguiTransformer.transform({ src, filename, options });
  } else if (filename.endsWith('.svg')) {
    return svgTransformer.transform({ src, filename, options });
  } else {
    return upstreamTransformer.transform({ src, filename, options });
  }
};
```

`metro.config.js`:

```js
const { getDefaultConfig, mergeConfig } = require('@react-native/metro-config');

const defaultConfig = getDefaultConfig(__dirname);
const { assetExts, sourceExts } = defaultConfig.resolver;

metroConfig = (() => {
  return {
    transformer: {
      babelTransformerPath: require.resolve('./customTransformer.js')
    },
    resolver: {
      assetExts: [...assetExts.filter((ext) => ext !== 'svg')],
      sourceExts: [...sourceExts, 'svg', 'po', 'pot']
    }
  };
})();

module.exports = mergeConfig(defaultConfig, metroConfig);

```
